### PR TITLE
refactor lstm_rewriter and support rewriting LSTMBlockCell

### PR DESCRIPTION
**Pull request by lucienwang1009**
Sorry for the big change. 

Previous lstm_rewriter (also gru_rewriter and grublock_rewriter) adopts a different logic from loop_rewriter.
The biggest difference is how to find the state variables of `dynamic_rnn` further served as inputs of `LSTM` op. The prior unit RNN rewriters find these state variables by matching nodes according to their name scope, which is less robust than the structure detection used in loop_rewriter.  **Thus, the new rewriters are based on loop_rewriter and fully utilize the loop attributes, such as state variables, already extracted by loop_rewriter.**

All unittest and internal tests of LSTM are passed and all `dynamic_rnn` with only one LSTM or LSTMBlockCell node are converted into onnx LSTM op.

### CODE REVIEW

File: tf2onnx/rewriter/unit_rnn_rewriter_base.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        super(UnitRnnContext, self).__init__()
+        self.rnn_scope = None
+        self.cell_match = None  # matched cell
+        self.rnn_props = RnnProperties()
+
+
+class UnitRnnRewriterBase(LoopRewriterBase):
+    def __init__(self, g, unittype):
+        super(UnitRnnRewriterBase, self).__init__(g)
+        self._unittype = unittype
```
Comment (nbcsm) [inline tf2onnx/rewriter/unit_rnn_rewriter_base.py]: required? seems only used for logging?

File: tf2onnx/rewriter/custom_rnn_rewriter.py [hunk 2]
```diff
@@ -39,63 +39,22 @@ def run(self):
...
+        res = parse_rnn_loop(self.g, context.loop_properties, context.rnn_scope,
+                             context.while_context_scope)
+        if not res:
             log.debug("skip the loop due to parse_rnn_loop failed")
             return False
 
-        self._parse_time_var(context)
+        time_var, iteration_var = res
+        context.time_var = time_var
+        context.iteration_var = iteration_var
```
Comment (pengwa) [inline tf2onnx/rewriter/custom_rnn_rewriter.py]: seems both time_var and iteration_var are useless here, if no more reference, just remove them.

File: tf2onnx/rewriter/loop_rewriter_base.py [hunk 3]
```diff
@@ -47,8 +48,8 @@ def __init__(self):
         # use enter name as key, they are initial inputs.
         # we don't use enter_input_id because it might be
         # used as initial input for more than one Enter nodes.
-        self.state_variables = {}
-        self.scan_variables = {}
+        self.state_variables = OrderedDict()
```
Comment (pengwa) [inline tf2onnx/rewriter/loop_rewriter_base.py]: thanks for the fixing.!!

File: tf2onnx/rewriter/lstm_rewriter_v2.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            return None
+
+        return RnnWeights(w, b, ft)
+
+    def get_attributes(self, context):
+        if self.lstm_cell_type == RNNUnitType.LSTMBlockCell:
+            lstm_block_cell = context.cell_match.get_op("lstm_block_cell")
+            clip = float(lstm_block_cell.get_attr("cell_clip").f)
+            # current LSTM op cannot handle clip
+            if clip > 0:
```
Comment (pengwa) [inline tf2onnx/rewriter/lstm_rewriter_v2.py]: log a warning maybe?

File: tf2onnx/rewriter/rnn_utils.py [hunk 5]
```diff
@@ -40,37 +40,26 @@ def __init__(self, kernel, bias, forget_bias):
...
-        else:
-            self.c_init_input_id = c_init
-            self.h_init_input_id = h_init
-            self.share_init_node = False
-
-
 class RnnProperties:
     def __init__(self):
         # RNN input who are outside of rnn scope
+        self.input_x = None  # InputTensorArray
```
Comment (pengwa) [inline tf2onnx/rewriter/rnn_utils.py]: if you would like, you can just give up using RnnProperties, just define UnitRnnContext inheriting Context, and put all those properties in the context. Then next step, we can just remove this data structure along with old lstm rewriter code.

File: tf2onnx/rewriter/lstm_rewriter_v2.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+            raise ValueError("no initializers, unexpected")
+        assert init_h_id and init_c_id
+        rnn_props.onnx_input_ids["initial_h"] = init_h_id
+        rnn_props.onnx_input_ids["initial_c"] = init_c_id
+
+    def _process_non_tuple_ch_init_nodes(self, rnn_props):
+        input_id = rnn_props.state_variables["ct_ht"].enter_input_id
+        hidden_size = rnn_props.hidden_size
+
+        # todo: remove this once Fill ops is supported
```
Comment (pengwa) [inline tf2onnx/rewriter/lstm_rewriter_v2.py]: maybe it's time to do this todo now.

File: tf2onnx/rewriter/rnn_utils.py [hunk 7]
```diff
@@ -40,37 +40,26 @@ def __init__(self, kernel, bias, forget_bias):
...
+        self.state_variables = {}
+        self.state_is_tuple = True
         self.input_node = None
         self.input_id = None
         self.var_initializers = {}
 
-        self.onnx_input_ids = {}
-
+        self.attributes = {}
+        self.onnx_input_ids = {}  # onnx inputs: [X, W, R, B, sequence_lens, initial_h, initial_c, P]
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/rewriter/rnn_utils.py]: our logic doesn't depend on their order.

File: tf2onnx/rewriter/unit_rnn_rewriter_base.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                    return False
+            return True
+
+        state_variables = context.loop_properties.get_variable(checker)
+        if len(state_variables) != 1:
+            log.debug("found %d state variables", len(state_variables))
+            return None
+        return state_variables[0]
+
+    def _workaround_fill_ch_init_node(self, initializer_input_id, rnn_props):
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/rewriter/unit_rnn_rewriter_base.py]: sure.

File: tf2onnx/rewriter/unit_rnn_rewriter_base.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+class UnitRnnContext(Context):
+    def __init__(self):
+        super(UnitRnnContext, self).__init__()
+        self.rnn_scope = None
+        self.cell_match = None  # matched cell
+        self.rnn_props = RnnProperties()
+
+
+class UnitRnnRewriterBase(LoopRewriterBase):
+    def __init__(self, g):
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/rewriter/unit_rnn_rewriter_base.py]: sure, let me copy the comments from unit_rewriter_base

File: tf2onnx/rewriter/custom_rnn_rewriter.py (hunk 2 above)
Comment (lucienwang1009) [reply tf2onnx/rewriter/custom_rnn_rewriter.py]: hmm, they are used by old rewriter. remove them later when all changes about rewriter are checked in?

File: tf2onnx/rewriter/rnn_utils.py (hunk 5 above)
Comment (lucienwang1009) [reply tf2onnx/rewriter/rnn_utils.py]: sure!

File: tf2onnx/rewriter/custom_rnn_rewriter.py (hunk 2 above)
Comment (lucienwang1009) [reply tf2onnx/rewriter/custom_rnn_rewriter.py]: I now put variables used in new rewriter into UnitRnnContext. the RnnProperty will keep intact.

**Pull request merged**